tests/sys_crypto/tests-crypto-modes-ccms.c

```diff
@@ -0,0 +1,318 @@
+    int len, err, cmp;
+    size_t len_encoding = nonce_and_len_encoding_size - nonce_len;
+
+    TEST_ASSERT_MESSAGE(sizeof(data) >= output_expected_len,
+                        "Output buffer too small");
+
+    err = cipher_init(&cipher, CIPHER_AES_128, key, key_len);
+    TEST_ASSERT_EQUAL_INT(1, err);
+
+    len = cipher_encrypt_ccms(&cipher, adata, adata_len,
+                              mac_length, len_encoding,
+                              nonce, nonce_len, plain, plain_len, data);
+    TEST_ASSERT_MESSAGE(len >= 0, "Encryption failed");
+
+    TEST_ASSERT_EQUAL_INT(output_expected_len, len);
+    cmp = compare(output_expected, data, len);
+    TEST_ASSERT_MESSAGE(1 == cmp, "wrong ciphertext");
+}
+
+static void test_decrypt_op(const uint8_t *key, uint8_t key_len,
+                            const uint8_t *adata, size_t adata_len,
+                            const uint8_t *nonce, uint8_t nonce_len,
+                            const uint8_t *encrypted, size_t encrypted_len,
+                            const uint8_t *output_expected,
+                            size_t output_expected_len,
+                            uint8_t mac_length)
+{
+    cipher_t cipher;
+    int len, err, cmp;
+    size_t len_encoding = nonce_and_len_encoding_size - nonce_len;
+
+    TEST_ASSERT_MESSAGE(sizeof(data) >= output_expected_len,
+                        "Output buffer too small");
+
+    err = cipher_init(&cipher, CIPHER_AES_128, key, key_len);
+    TEST_ASSERT_EQUAL_INT(1, err);
+
+    len = cipher_decrypt_ccms(&cipher, adata, adata_len,
+                              mac_length, len_encoding,
+                              nonce, nonce_len, encrypted, encrypted_len, data);
+    TEST_ASSERT_MESSAGE(len >= 0, "Decryption failed");
+
+    TEST_ASSERT_EQUAL_INT(output_expected_len, len);
+    cmp = compare(output_expected, data, len);
+    TEST_ASSERT_MESSAGE(1 == cmp, "wrong ciphertext");
+
+}
+
+#define do_test_encrypt_op(name) do { \
+        test_encrypt_op(TEST_ ## name ## _KEY, TEST_ ## name ## _KEY_LEN, \
+                        TEST_ ## name ## _INPUT, TEST_ ## name ## _ADATA_LEN, \
+                        TEST_ ## name ## _NONCE, TEST_ ## name ## _NONCE_LEN, \
+                    \
+                        TEST_ ## name ## _INPUT + TEST_ ## name ## _ADATA_LEN, \
+                        TEST_ ## name ## _INPUT_LEN, \
+                    \
+                        TEST_ ## name ## _EXPECTED + TEST_ ## name ## _ADATA_LEN, \
+                        TEST_ ## name ## _EXPECTED_LEN - TEST_ ## name ## _ADATA_LEN, \
+                    \
+                        TEST_ ## name ## _MAC_LEN \
+                        ); \
+} while (0)
+
+
+static void test_crypto_modes_ccms_encrypt(void)
+{
+    do_test_encrypt_op(IEEE_0537R2_1);
+    do_test_encrypt_op(IEEE_0537R2_2);
+    do_test_encrypt_op(IEEE_0537R2_3);
+}
+
+#define do_test_decrypt_op(name) do { \
+        test_decrypt_op(TEST_ ## name ## _KEY, TEST_ ## name ## _KEY_LEN, \
```

Comment:
I'm not familiar with the new *tests/sys_crypto* structure. Is it possible to share this between *tests-crypto-modes-ccm.c* and *tests-crypto-modes-ccms.c* ?

Reply:
The macros could be in a shared header file yes, and I could move the shared `test_decrypt_op()/test_encrypt_op` into a helper source file as well.

Reply:
Ah, this hasn't been addressed yet.